modules/utask/execution.go

```diff
@@ -0,0 +1,141 @@
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package utask
+
+import (
+	"encoding/gob"
+	"fmt"
+	"reflect"
+	"runtime"
+
+	"github.com/pkg/errors"
+)
+
+var fnNameMap = make(map[string]interface{})
+
+// Signature represents a function and its arguments
+type Signature struct {
+	FnName string
+	Args   []interface{}
+}
+
+// Execute executes the function
+func (s *Signature) Execute() ([]reflect.Value, error) {
+	var targetArgs []reflect.Value
+	for _, arg := range s.Args {
+		targetArgs = append(targetArgs, reflect.ValueOf(arg))
+	}
+	if fn, ok := fnNameMap[s.FnName]; ok {
+		fnValue := reflect.ValueOf(fn)
+		return fnValue.Call(targetArgs), nil
+	}
+	return nil, fmt.Errorf("Function: %s not found. Did you forget to register?", s.FnName)
+}
+
+// Register registers a function for async tasks
+func Register(fn interface{}) error {
+	fnType := reflect.TypeOf(fn)
+	if err := validateAsyncFn(fnType); err != nil {
+		return err
+	}
+	for i := 0; i < fnType.NumIn(); i++ {
+		gob.Register(reflect.Zero(fnType.In(i)).Interface())
+	}
+	fnName := getFunctionName(fn)
+	fnNameMap[fnName] = fn
+	return nil
+}
+
+// Enqueue sends a func before sending to the task queue
+func Enqueue(fn interface{}, args ...interface{}) error {
+	fnType := reflect.TypeOf(fn)
+	if fnType.NumIn() != len(args) {
+		return fmt.Errorf("Expected %d function arg(s) but found %d\n", fnType.NumIn(), len(args))
+	}
+	var argValues []reflect.Value
+	for i := 0; i < fnType.NumIn(); i++ {
+		arg := reflect.ValueOf(args[i])
+		argType := reflect.TypeOf(args[i])
+		if !argType.AssignableTo(fnType.In(i)) {
+			// TODO(madhu): Is it useful to show the arg index or the arg value in the error msg?
+			return fmt.Errorf(
+				"Cannot assign function argument: %d from type: %s to type: %s\n",
+				i+1, argType, fnType.In(i),
+			)
+		}
+		argValues = append(argValues, arg)
+	}
+	fnName := getFunctionName(fn)
+	s := Signature{FnName: fnName, Args: args}
+
+	sBytes, err := GlobalBackend().Encoder().Marshal(s)
+	if err != nil {
+		return errors.Wrap(err, "unable to encode the function")
```

Comment:
Can't say at this point, depends on what the error is, can clarify to say function or args.